src/cpp/src/sequence_group.hpp

```diff
@@ -212,6 +217,73 @@ class SequenceGroup {
         m_sequences.erase(remove_it);
     }
 
+    // Called when sampling params contain stop_strings
+    void include_tokenizer(Tokenizer& tokenizer) {
+        m_tokenizer = tokenizer;
+    }
+
+    // Handle stop_strings
+    bool stop_string_hit(const TokenIds & generated_tokens) {
+        /*
+        For catching stop_string hit we run comparisons character-wise to catch cases where stop string 
+        overlaps with part of another token on both sides or is just a part of a single token. 
+        For every stop_string we iterate over generated tokens starting from the last one and going backwards. 
+        Every token is decoded and its characters are compared to the stop_string character at a current_position 
+        (position of a character in the stop_string counting from the last one) - at the begining position is 0.
+        When characters match we increase current_position and check if we have a full match already, if not we continue.
+        If we have already matched some characters (current_position > 0) and next character is not matching 
+        before we reach the full match, then we reset current_position to 0. 
+        */
+        for (auto stop_string: m_sampling_params.stop_strings) {
+            int current_position = 0;
+            // Getting reverse iterator to check tokens starting from the last one generated and going backwards
+            auto generated_tokens_rit = generated_tokens.rbegin();
+            while (generated_tokens_rit != generated_tokens.rend()) {
+                std::string decoded_token = m_tokenizer.decode(std::vector<int64_t>{*generated_tokens_rit});
+                // Checking decoded_token characters starting from the last one
+                for (auto decoded_token_rit = decoded_token.rbegin(); decoded_token_rit != decoded_token.rend(); decoded_token_rit++) {
+                    // On character match increment current_position for the next comparisons
+                    if (*decoded_token_rit == *(stop_string.rbegin() + current_position)) {
+                        current_position++;
+                        // If this is the last character from the stop_string we have a match
+                        if ((stop_string.rbegin() + current_position) == stop_string.rend()) {
+                            return true;
+                        } 
+                    } else if (current_position) {
+                        // Already found matching characters, but the last one didn't match, so we reset current_position
+                        current_position = 0;
+                    }
+                }
+                generated_tokens_rit++;
+            }
+        }
+        return false;
+    }
+
+    // Handle stop_token_ids
+    bool stop_token_ids_hit(const TokenIds & generated_tokens) {
```

Comment:
on code-wise level it does not look good to me, that sampling responsibility is now part of sequence group. Sequence group is supposed to manage sequence state.

Maybe we can move all these new methods, including `try_finish_generation`, to sampler's code? `try_finish_generation` could be just a free function, that accepts sequence group as argument.